symbolic-debuginfo/src/dwarf.rs

```diff
@@ -829,182 +768,228 @@ impl<'d, 'a> DwarfUnit<'d, 'a> {
-                                let line_info = LineInfo {
-                                    address: range_begin,
-                                    size: Some(range_end.min(next.address) - range_begin),
-                                    file: file.clone(),
-                                    line,
-                                };
-
-                                lines.insert(index, line_info);
-                                index += 1;
-                            }
-                        }
-
-                        while index < lines.len() {
-                            let record = &mut lines[index];
+            seen_ranges.insert((address, size))
+        });
 
-                            // Advance to the next range since we're done here.
-                            if record.address >= range_end {
-                                break;
-                            }
+        // Ranges can be empty for three reasons: (1) the function is a no-op and does not
+        // contain any code, (2) the function did contain eliminated dead code, or (3) some
+        // tooling created garbage reversed ranges which we filtered out.
+        // In the dead code case, a surrogate DIE remains with `DW_AT_low_pc(0)` and empty ranges.
+        // That DIE might still contain inlined functions with actual ranges - these must be skipped.
+        // However, non-inlined functions may be present in this subtree, so we must still descend
+        // into it.
+        if ranges.is_empty() {
+            return self.parse_functions(depth, entries, output);
+        }
 
-                            index += 1;
+        // Resolve functions in the symbol table first. Only if there is no entry, fall back
+        // to debug information only if there is no match. Sometimes, debug info contains a
+        // lesser quality of symbol names.
+        //
+        // XXX: Maybe we should actually parse the ranges in the resolve function and always
+        // look at the symbol table based on the start of the DIE range.
+        let symbol_name = if self.prefer_dwarf_names {
+            None
+        } else {
+            let first_range_begin = ranges.iter().map(|range| range.begin).min().unwrap();
+            let function_address = offset(first_range_begin, self.inner.info.address_offset);
+            self.resolve_symbol_name(function_address)
+        };
 
-                            // Skip forward to the next line record that overlaps with our range.
-                            // Lines before belong to the parent function or another inlinee.
-                            let record_end = record.address + record.size.unwrap_or(0);
-                            if record_end <= range_begin {
-                                continue;
-                            }
+        let name = symbol_name
+            .or_else(|| self.resolve_dwarf_name(&self.inner.unit.entry(dw_die_offset).unwrap()))
```

Comment:
can the `dw_die_offset` ever be invalid? I think so, given we parse that from the raw entries without validation.
In that case I think we should raise a proper Err here, and down below for inlinees.

Reply:
The `dw_die_offset` is just the byte offset of the reader, as far as I can tell. It comes from the caller who just parsed an abbrev from that very offset. So I think it's always valid.

Reply:
So what happens here is that we parse the attributes for the same entry twice: Once to get the address ranges, and then again to get the name. There is an opportunity here to merge the two, but I didn't want to bother with that change in a PR that's already quite large. So the call to `unit.entry()` is more of an escape hatch.